api/pull.go

```diff
@@ -131,28 +154,65 @@ func (hs historyServiceResponder) getMesosAgents(jsonPath []string) (nodes []Nod
 		return nodes, errors.New("Agent nodes were not found in hisotry service for the past hour")
 	}
 	for ip, _ := range nodeCount {
-		var node Node
-		node.Role = "agent"
-		node.Ip = ip
-		nodes = append(nodes, node)
+		nodes = append(nodes, Node{
+			Role: ROLE_AGENT,
+			Ip:   ip,
+		})
 	}
 	return nodes, nil
 }
 
-func (dr dnsResponder) getAgentSource() (leaderIps []string, err error) {
-	leaderIps, err = net.LookupHost(dr.defaultMasterAddress)
+func (f *findAgentsInHistoryService) find() (nodes []Node, err error) {
+	nodes, err = f.getMesosAgents()
+	if err == nil {
+		return nodes, nil
+	}
+	// try next provider if it is available
+	if f.next != nil {
+		log.Warning(err)
+		return f.next.find()
+	}
+	return nodes, err
+}
+
+// find agents by resolving dns entry
+type findNodesInDns struct {
+	dnsRecord string
+	role      string
+	next      nodeFinder
+}
+
+func (f *findNodesInDns) resolveDomain() (ips []string, err error) {
```

Comment:
Also, `Dns` should be `DNS` per Go initialisms constructs: https://github.com/golang/go/wiki/CodeReviewComments#initialisms